src/mc-textopts.c

```diff
@@ -0,0 +1,305 @@
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMinQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMinQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(strMaxQueryLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMaxQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        CLIENT_ERR(ERROR_PREFIX "Unrecognized field: '%s'", field);
+        return false;
+    }
+    return true;
+}
+
+bool mc_TextOpts_parse(mc_TextOpts_t *txo, const bson_t *in, mongocrypt_status_t *status) {
+    bson_iter_t iter = {0};
+    BSON_ASSERT_PARAM(txo);
+    BSON_ASSERT_PARAM(in);
+    BSON_ASSERT(status || true);
+    bool has_caseSensitive = false, has_diacriticSensitive = false, has_substring = false, has_prefix = false,
+         has_suffix = false;
+
+    *txo = (mc_TextOpts_t){0};
+    txo->bson = bson_copy(in);
+
+    if (!bson_iter_init(&iter, txo->bson)) {
+        CLIENT_ERR(ERROR_PREFIX "Invalid BSON");
+        return false;
+    }
+    while (bson_iter_next(&iter)) {
+        const char *field = bson_iter_key(&iter);
+        IF_FIELD(caseSensitive);
+        {
+            if (!BSON_ITER_HOLDS_BOOL(&iter)) {
+                CLIENT_ERR(ERROR_PREFIX "Expected bool for caseSensitive, got: %s",
+                           mc_bson_type_to_string(bson_iter_type(&iter)));
+                return false;
+            }
+            txo->caseSensitive = bson_iter_bool(&iter);
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(diacriticSensitive);
+        {
+            if (!BSON_ITER_HOLDS_BOOL(&iter)) {
+                CLIENT_ERR(ERROR_PREFIX "Expected bool for diacriticSensitive, got: %s",
+                           mc_bson_type_to_string(bson_iter_type(&iter)));
+                return false;
+            }
+            txo->diacriticSensitive = bson_iter_bool(&iter);
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(substring);
+        {
+            bson_iter_t subdoc;
+            if (!BSON_ITER_HOLDS_DOCUMENT(&iter) || !bson_iter_recurse(&iter, &subdoc)) {
+                CLIENT_ERR(ERROR_PREFIX "Expected document for substring, got: %s",
+                           mc_bson_type_to_string(bson_iter_type(&iter)));
+                return false;
+            }
+
+            if (!mc_TextOptsPerIndex_parse(&txo->substring, &subdoc, status)) {
+                return false;
+            }
```

Comment:
Suggest erroring here if `strMaxLength` is not set for substring.

```c
if (!txo->substring.strMaxLength.set) {
    CLIENT_ERR(ERROR_PREFIX "'strMaxLength' must be set for substring");
    return false;
}
```

Without this check, omitting `strMaxLength` produces an error message that is likely less useful to users (referencing payloads, rather than options):
```
Error parsing FLE2SubstringInsertSpec: Missing field 'mlen' in placeholder
```